### ISSUE TITLE
BUG: Method `duplicated` performances on geometry column

### ISSUE THREAD

**Issue by benoit9126** [geopandas/geopandas#2607]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of geopandas.

- [ ] (optional) I have confirmed this bug exists on the main branch of geopandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

I would like to report a performance problem.

```python
import timeit

stmt = """gdf["geometry"].duplicated(keep=False)"""
setup = """
import geopandas as gpd
from shapely.geometry import LineString

size={size}
gdf=gpd.GeoDataFrame(data={{"a":[1]*size, "b":[3]*size}}, geometry=[LineString([(0,0), (1,1)])]*size)
"""

# Duplicated applied on a single geometry column for different size of GeoDataFrame -> slow
timeit.timeit(stmt=stmt, setup=setup.format(size=10), number=10)
# 0.023582907999752933
timeit.timeit(stmt=stmt, setup=setup.format(size=100), number=10)
# 0.08639520700035064
timeit.timeit(stmt=stmt, setup=setup.format(size=1000), number=10)
# 0.5222119659997588
timeit.timeit(stmt=stmt, setup=setup.format(size=10000), number=10)
# 5.11025889100074

# Just to compare with a non-geometry columns (much quicker)
stmt = """gdf["a"].duplicated(keep=False)"""
timeit.timeit(stmt=stmt, setup=setup.format(size=10), number=10)
# 0.0014762759997211106
timeit.timeit(stmt=stmt, setup=setup.format(size=100), number=10)
# 0.0014409380000870442
timeit.timeit(stmt=stmt, setup=setup.format(size=1000), number=10)
# 0.0010032489999503014
timeit.timeit(stmt=stmt, setup=setup.format(size=10000), number=10)
# 0.0012840540002798662

# If now, I try to look for duplicates on two columns (including the geometry) -> quick again (not the same algorithm in Pandas)
stmt = """gdf[["geometry", "a"]].duplicated(keep=False)"""
timeit.timeit(stmt=stmt, setup=setup.format(size=10), number=10)
# 0.03693144399994708
timeit.timeit(stmt=stmt, setup=setup.format(size=100), number=10)
# 0.030359643999872787
timeit.timeit(stmt=stmt, setup=setup.format(size=1000), number=10)
# 0.034004715000264696
timeit.timeit(stmt=stmt, setup=setup.format(size=10000), number=10)
# 0.08402770300017437

# I can even do it with two geometry columns and it is quick as expected
setup = """
import geopandas as gpd
from shapely.geometry import LineString

size={size}
gdf=gpd.GeoDataFrame(data={{"a":[1]*size, "b":[3]*size}}, geometry=[LineString([(0,0), (1,1)])]*size)
gdf["geometry2"] = gdf["geometry"]
"""
stmt = """gdf[["geometry", "geometry2"]].duplicated(keep=False)"""
timeit.timeit(stmt=stmt, setup=setup.format(size=10), number=10)
# 0.03633627599992906
timeit.timeit(stmt=stmt, setup=setup.format(size=100), number=10)
# 0.04178242199986926
timeit.timeit(stmt=stmt, setup=setup.format(size=1000), number=10)
# 0.04884332900019217
timeit.timeit(stmt=stmt, setup=setup.format(size=10000), number=10)
# 0.15179994200025249
```

#### Problem description

When I am looking for duplicated geometries, the `duplicated` method of pandas is extremely slow when I only provide a single column (of geometries). If I provide two columns (at least) which can be both geometries or geometry + another type of column, the `duplicated` is much faster.

I sometimes have geo data frame of several thousand rows and the `duplicated` "never ends" (took several hours when I provide a single column and only a few minutes with two duplicated columns of geometries)

#### Expected Output

I would have expected to have similar performance between `duplicated` method applied on a single geometry column and two columns (geometry or not).

In Pandas, if there is a single column, they use a hash table implemented in `_libs.hashtable`. The separation between single column duplication and multi-column duplication is done [here](https://github.com/pandas-dev/pandas/blob/890d09753492242bd30cc43c7ebd4e819ad89bf1/pandas/core/frame.py#L6817).

It may be interesting to have a `duplicated` method in `GeoSeries` which would force using the pandas algorithm used for two (or more) column duplication.

I realized that they have a hash table per data type (`Float32HashTable`, `Float64HashTable`, ...). In a future release, it may be interesting to implement a hash table for geometries using the shapely 2.0 hash method of geometries objects.

#### Output of ``geopandas.show_versions()``

<details>

SYSTEM INFO
-----------

python     : 3.9.6 (default, Aug 23 2021, 19:59:27)  [GCC 10.2.1 20210110]
executable : /home/benoit/.pyenv/versions/sirao_cli/bin/python
machine    : Linux-5.10.0-19-amd64-x86_64-with-glibc2.31

GEOS, GDAL, PROJ INFO
---------------------

GEOS       : 3.9.0
GEOS lib   : /usr/lib/x86_64-linux-gnu/libgeos_c.so
GDAL       : 3.4.3
GDAL data dir: /home/benoit/.pyenv/versions/sirao_cli/lib/python3.9/site-packages/fiona/gdal_data
PROJ       : 9.1.0
PROJ data dir: /home/benoit/.pyenv/versions/sirao_cli/lib/python3.9/site-packages/pyproj/proj_dir/share/proj

PYTHON DEPENDENCIES
-------------------

geopandas  : 0.11.1
numpy      : 1.23.4
pandas     : 1.5.1
pyproj     : 3.4.0
shapely    : 1.8.5.post1
fiona      : 1.8.22
geoalchemy2: 0.12.5
geopy      : 2.2.0
matplotlib : 3.6.1
mapclassify: None
pygeos     : 0.13
pyogrio    : None
psycopg2   : None
pyarrow    : None
rtree      : 1.0.1

</details>

**Comment by jorisvandenbossche** [comment]
@benoit9126 thanks for the report! 
I can confirm this, and from taking a quick look, it seems most of the slow performance can be attributed to the pygeos->shapely conversion. So what happens under the hood is that pandas is converting the geometry column to a numpy array `np.array(gdf["geometry"]`). And if you have PyGEOS installed and enabled (by default), then in this conversion to numpy array, we do convert the pygeos geometries to shapely geometries (that was done to preserve backwards compatibility). 

So that's an unfortunate side effect of using PyGEOS (which generally gives speedups when installed, not slowdowns ...)

A possible workaround with the currently released versions is to first convert to WKB:

```
In [3]: %timeit gdf["geometry"].duplicated(keep=False)
425 ms ± 27.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [4]: %timeit gdf["geometry"].to_wkb().duplicated(keep=False)
6.39 ms ± 134 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Also to note that with Shapely 2.0 (beta release is out and can be tested, although right now you need geopandas dev version), this conversion doesn't happen anymore, and I can confirm a speed up in that case as well (without the explicit conversion to WKB):

```
# with USE_PYGEOS=0 env variable set before starting IPython
In [2]: %timeit gdf["geometry"].duplicated(keep=False)
7.61 ms ± 63.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by benoit9126** [comment]
Thanks for the answer @jorisvandenbossche. I will convert my columns into WKB as you suggested for the moment. I am eager to see Shapely 2.0+GeoPandas compatible version released. ^^

**Comment by martinfleis** [comment]
I will close this as it there is no action to be taken. This will gets resolved by switching to shapely 2.0 backend.
